inspirehep/records/api/base.py

```diff
@@ -166,3 +179,228 @@ def hard_delete(self):
+            record_id = self.id
+        try:
+            location_obj = Location.get_by_name(location)
+        except NoResultFound:
+            raise NoResultFound(
+                f"Cannot find location '{location}'. "
+                f"Please check if system is configured properly!"
+            )
+
+        bucket = (
+            RecordsBuckets.query.join(Bucket)
+            .filter(
+                RecordsBuckets.record_id == record_id,
+                Bucket.default_storage_class == storage_class,
+                Bucket.default_location == location_obj.id,
+            )
+            .one_or_none()
+        )
+        if bucket:
+            return bucket.bucket
+        return self._create_bucket(location, storage_class)
+
+    def _create_bucket(self, location=None, storage_class=None):
+        """Create bucket and return it.
+        Note:
+            Overwrites base_class._create_bucket method as it is not implemented
+            It can create more than one bucket for the same parameters.
+            It's private method, do not use it. Instead use `get_bucket()`
+
+        Args:
+            location (Location): Bucket location object
+                (default: 'RECORDS_DEFAULT_FILE_LOCATION_NAME') from config
+            storage_class (str): Bucket storage class
+                (default: 'RECORDS_DEFAULT_STORAGE_CLASS') from config
+
+        Returns: Bucket for current record, selected location and storage_class
+        """
+        bucket = Bucket.create(location=location, storage_class=storage_class)
+        RecordsBuckets.create(record=self.model, bucket=bucket)
+        return bucket
+
+    @property
+    def files(self):
+        """Get files iterator.
+        Original one was so ugly that it had to be changed!
+
+        Returns:
+            Iterator to files
+        """
+        if self.model is None:
+            raise MissingModelError()
+
+        bucket = self.get_bucket()
+
+        return self.files_iter_cls(self, bucket=bucket, file_cls=self.file_cls)
+
+    def _download_file_from_url(self, url):
+        """Downloads file and calculates hash for it
+        if everythong is ok then adds it to files in current record
+
+        Args:
+            url (str): Local or remote url/filepath
+
+        Returns (str): key(sha-1) of downloaded file
+
+        Example:
+            >>> self._download_file_from_url('http://example.com/url_to_file.pdf')
+            '207611e7bf8a83f0739bb2e16a1a7cf0d585fb5f'
+        """
+        stream = fsopen(url, mode="rb")
+        # TODO: change to stream.read() when fs will be updated to > 2.0
+        # As HTTPOpener is not working with size = -1
+        # (and read() method sets this size as default)
+        # This is workaround until we will update to fs >2.0
+        data = stream._f.wrapped_file.read()
+        key = hashlib.sha1(data).hexdigest()
+        if key not in self.files.keys:
+            self.files[key] = BytesIO(data)
+        return key
```

Comment:
what will happen if the download fails? It's better to through a custom error

Reply:
If download fails we won't be in this part of the code. Exception `ResourceNotFoundError` will be thrown and in `_find_and_add_file` empty key will be returned which won't allow to add this file to documents but won't break other files in the list. Later on in workflow or other place  which will be calling add_files or set_files there should be check if all files was added correctly. May be I should look again in `create` and `update` methods to somehow inform about missing files.